src/gpgmm/common/CPUPoolAllocator.h

```diff
@@ -0,0 +1,93 @@
+// Copyright 2022 The GPGMM Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GPGMM_COMMON_CPUPOOLALLOCATOR_H_
+#define GPGMM_COMMON_CPUPOOLALLOCATOR_H_
+
+#include "gpgmm/utils/Assert.h"
+
+#include <queue>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+using AllocationAddress = std::pair<uint64_t, uint64_t>;
+
+namespace gpgmm {
+    template <typename T>
+    class CPUPoolAllocator {
+      public:
+        explicit CPUPoolAllocator(size_t blockSize, size_t objectSize) {
+            mBlockSize = blockSize;
+            mObjectsPerBlock = blockSize / objectSize;
+        }
+
+        T* Allocate(T object) {
+            // If a free allocation is available, reuse that location.
+            if (!mFreeQueue.empty()) {
+                AllocationAddress freeSpot = mFreeQueue.front();
+                mFreeQueue.pop();
+                mAllocations[freeSpot.first][freeSpot.second] = object;
+                return &mAllocations[freeSpot.first][freeSpot.second];
+            }
+
+            // If our next allocation offset starts with a zero it means we haven't allocated the
+            // block yet. Allocate it.
+            if (mNextAllocation.second == 0) {
+                mAllocations.push_back(std::vector<T>());
+                mAllocations[mNextAllocation.first].reserve(mObjectsPerBlock);
+            }
+
+            mAllocations[mNextAllocation.first].push_back(object);
+
+            T* allocationAddress = &mAllocations[mNextAllocation.first][mNextAllocation.second];
+
+            // Map the pointer address to the AllocationAddress so we can keep track of this
+            // location for deallocation.
+            mAllocationAddressMap.insert(
+                {allocationAddress, {mNextAllocation.first, mNextAllocation.second}});
+
+            // Increment next allocation position.
+            mNextAllocation.second++;
+            if (mNextAllocation.second == mObjectsPerBlock) {
+                mNextAllocation.first++;
+                mNextAllocation.second = 0;
+            }
+
+            return allocationAddress;
+        }
+
+        void Deallocate(T* objectPointer) {
+            // Translate T* pointer back to an AllocationAddress
+            typename std::unordered_map<T*, AllocationAddress>::const_iterator search =
```

Comment:
Why not store AllocationAddress in some region outside of T*? No `std::unordered_map` required.